common/quotas/global/collection/collection.go

```diff
@@ -463,15 +464,18 @@ func (c *Collection) doUpdate(since time.Duration, usage map[shared.GlobalKey]rp
 			continue
 		}
 
-		if weight < 0 {
+		if info.Weight < 0 {
 			// negative values cannot be valid, so they're a failure.
 			//
 			// this is largely for future-proofing and to cover all possibilities,
 			// so unrecognized values lead to fallback behavior because they cannot be understood.
 			c.global.Load(lkey).FailedUpdate()
 		} else {
-			target := float64(c.targetRPS(lkey))
-			c.global.Load(lkey).Update(rate.Limit(weight * target))
+			target := rate.Limit(c.targetRPS(lkey))
+			limiter := c.global.Load(lkey)
+			fallbackTarget := limiter.FallbackLimit()
+			boosted := boostRPS(target, fallbackTarget, info.Weight, info.UsedRPS)
+			limiter.Update(boosted)
```

Comment:
everything else is in service of this change here: boost limits before applying them.